doozerlib/distgit.py

```diff
@@ -1411,23 +1433,45 @@ def update_distgit_dir(self, version, release, prev_release=None):
                 if in_mod_block:
                     continue
 
-                # If we are injecting environment variables, remove any line
-                # from the Dockerfile that may be trying to set the same name. This is also
-                # necessary for us to clean up our older env injections if this is a dist-git only
-                # Dockerfile.
+                # If we are injecting environment variables, mangle
+                # environment variables with the same name so that doozer has the last say.
                 if line.lower().startswith('env '):
-                    if any(f' {var_name}=' in line for var_name in env_vars):
+
+                    # IF this line is a set of environment variables introduced by updates,
+                    # discard the line. We will inject the vars later.
+                    # The check for OS_GIT_MAJOR cleans up pre-flag doozer injected vars.
+                    if env_update_line_flag in line or 'OS_GIT_MAJOR' in line:
+                        continue
+
+                    # If this line is a set of environment variables introduced from source
+                    # previously AND we have new items from source, discard the line entirely.
+                    # We will inject the new vars later.
+                    if self.env_vars_from_source is not None and env_merge_line_flag in line:
                         continue
 
+                    # Otherwise, this is an env line that doozer did not create.
+                    # Split:
+                    #     "env  a=1 b=2 c=3" into [ 'env', 'a=1', ...]
+                    # or  "env a 1" into ['env', 'a', '1' ]  . Dockerfiles support both forms.
+                    env_comp = line.split()
+                    for i, comp in enumerate(env_comp[1:], 1):
+                        if comp.split('=')[0] in env_vars:
+                            # Make the old value be ignored
+                            env_comp[i] = f'__overridden_{comp}'
+                    line = ' '.join(env_comp)   # Reformulate the line
```

Comment:
I worry just a little bit about reading and rewriting ENV statements; that's harder than just writing them. dockerfile_parse does have [some code dealing with envs](https://github.com/containerbuildsystem/dockerfile-parse/blob/a02edc4ee12a0900390bcae5d43a7df34a850a12/dockerfile_parse/parser.py#L407) but it's admittedly difficult to see if it's useful here (maybe the parsing part of it?)

Reply:
I've changed things up a bit  in the latest commit to use the dockerfileparser: https://github.com/openshift/doozer/pull/201/commits/62b6f6c8ec62fda5899960b3ea5faee8308b5542 . Had to overcome some of its limits, but I think you are right that it is superior.